pkg/cache/v3/snapshot.go

```
@@ -94,64 +36,32 @@ type Snapshot struct {
 	out := Snapshot{}
-	out.Resources[types.Endpoint] = NewResources(version, resources.Endpoints)
-	out.Resources[types.Cluster] = NewResources(version, resources.Clusters)
-	out.Resources[types.Route] = NewResources(version, resources.Routes)
-	out.Resources[types.Listener] = NewResources(version, resources.Listeners)
-	out.Resources[types.Runtime] = NewResources(version, resources.Runtimes)
-	out.Resources[types.Secret] = NewResources(version, resources.Secrets)
-	out.Resources[types.ExtensionConfig] = NewResources(version, resources.ExtensionConfigs)
+
+	for typ, resource := range resources {
+		index := GetResponseType(typ)
+		if index != types.UnknownType {
+			out.Resources[index] = NewResources(version, resource)
+		}
+	}
 
 	return out
 }
 
-func NewSnapshotWithTtls(version string,
-	endpoints []types.ResourceWithTTL,
-	clusters []types.ResourceWithTTL,
-	routes []types.ResourceWithTTL,
-	listeners []types.ResourceWithTTL,
-	runtimes []types.ResourceWithTTL,
-	secrets []types.ResourceWithTTL) Snapshot {
+// NewSnapshotWithTTLs creates a snapshot with time to live per resource.
+// The resources map is keyed off the type URL of a resource, followed by the slice of resource objects.
+func NewSnapshotWithTTLs(version string, resources map[resource.Type][]types.ResourceWithTTL) Snapshot {
 	out := Snapshot{}
-	out.Resources[types.Endpoint] = NewResourcesWithTtl(version, endpoints)
-	out.Resources[types.Cluster] = NewResourcesWithTtl(version, clusters)
-	out.Resources[types.Route] = NewResourcesWithTtl(version, routes)
-	out.Resources[types.Listener] = NewResourcesWithTtl(version, listeners)
-	out.Resources[types.Runtime] = NewResourcesWithTtl(version, runtimes)
-	out.Resources[types.Secret] = NewResourcesWithTtl(version, secrets)
+
+	for typ, resource := range resources {
+		index := GetResponseType(typ)
+		if index != types.UnknownType {
+			out.Resources[index] = NewResourcesWithTTL(version, resource)
+		}
```

If we hit this it seems like a programmer error, should we return an error?

---

I think an error here would make sense. Right now it's a silent ignore which might be confusing to users